modules/dnn/src/caffe/caffe_importer.cpp

```diff
@@ -334,12 +342,49 @@ class CaffeImporter
 
         //setup input layer names
         std::vector<String> netInputs(net.input_size());
+        std::vector<MatShape> inp_shapes;
         {
-            for (int inNum = 0; inNum < net.input_size(); inNum++)
+            int net_input_size = net.input_size();
+            for (int inNum = 0; inNum < net_input_size; inNum++)
             {
                 addedBlobs.push_back(BlobNote(net.input(inNum), 0, inNum));
                 netInputs[inNum] = net.input(inNum);
             }
+
+            if (net.input_dim_size() > 0)  // deprecated in Caffe proto
+            {
+                int net_input_dim_size = net.input_dim_size();
+                CV_Check(net_input_dim_size, net_input_dim_size % 4 == 0, "");
+                CV_CheckEQ(net_input_dim_size, net_input_size * 4, "");
```

Comment:
Maybe just just check that number of `input_dim` is divided by number of inputs?
```cpp
CV_Check(net_input_dim_size, net_input_dim_size % net_input_size == 0, "");
```
So we could handle multiple inputs but with different number of dimensions (considering that it is the same for each input).

We have an example with different number of shapes (`faster_rcnn_vgg16.prototxt`). Looks like that's why `input_dim` is deprecated in Caffe.
```yaml
input: "data"
input_shape {
  dim: 1
  dim: 3
  dim: 224
  dim: 224
}

input: "im_info"
input_shape {
  dim: 1
  dim: 3
}
```

Reply:
This code block is about "input_dim".
It is strictly restricted to 4 elements, see [proto file comment](https://github.com/opencv/opencv/blob/3.4.9/modules/dnn/src/caffe/opencv-caffe.proto#L227-L231).

"input_shape" doesn't have such limitations (related code is below).

Reply:
Got it, thanks!